### PR TITLE
Add k eval predictions to mlflow for LLM fine-tuning

### PR DESCRIPTION
**Pull request by Infernaught**
Adds LLM predictions to MLFlow during evaluation. Local behavior shows inputs, outputs, and prompts added to a csv file in the MLFlow logs.

### CODE REVIEW

File: ludwig/trainers/trainer.py [hunk 1]
```diff
@@ -659,6 +660,34 @@ def run_evaluation(
...
+                o_feat.feature_name: torch.from_numpy(
+                    np.array(training_set.dataset[o_feat.proc_column][:num_eval_examples], copy=True)
+                ).to(self.device)
+                for o_feat in self.dist_model.output_features.values()
+            }
+            eval_example_outputs = self.dist_model((inputs, targets))["predictions"]
+            llm_outputs = {"inputs": [], "outputs": []}
+            input_list = list(inputs.values())[0]
+            for i in range(len(eval_example_outputs)):
+                clean_input = remove_left_padding(input_list[i], self.dist_model.tokenizer)[0]
```
Comment (justinxzhao) [inline ludwig/trainers/trainer.py:682]: Instead of surfacing lower-level code and the raw forward pass, could we re-use `self.evaluation`, which could be modified to return predictions:

```python
    def evaluation(self, dataset, dataset_name, metrics_log, batch_size, progress_tracker):
        predictor = Predictor(
            self.dist_model,
            batch_size=batch_size,
            distributed=self.distributed,
            report_tqdm_to_ray=self.report_tqdm_to_ray,
            model=self.model,
        )
        ### Setting collect_predictions=True will return predictions.
        metrics, _ = predictor.batch_evaluation(dataset, collect_predictions=False, dataset_name=dataset_name)

        return append_metrics(self.model, dataset_name, metrics, metrics_log, progress_tracker)
```

File: ludwig/trainers/trainer_llm.py
Comment (Infernaught) [inline ludwig/trainers/trainer_llm.py:1]: This is still not a super lightweight solution, but it turns out that setting `collect_predictions=True` causes the code to break when doing batch evaluation because we cannot concatenate the outputs when they are different shapes.

Comment (justinxzhao) [reply ludwig/trainers/trainer_llm.py:1]: Yikes! That's good to know -- could you leave a comment in `self.evaluation()` to mention this?

File: ludwig/trainers/trainer_llm.py [hunk 2]
```diff
@@ -434,8 +435,28 @@ def evaluation(self, dataset, dataset_name, metrics_log, batch_size, progress_tr
...
+        llm_eval_examples = {"inputs": [], "targets": [], "outputs": []}
+        for key in inp_tar_out_dict["inputs"]:
+            for inp in inp_tar_out_dict["inputs"][key]:
+                inp = remove_left_padding(inp, tokenizer)[0]
+                llm_eval_examples["inputs"].append(tokenizer.decode(inp))
+
+        for key in inp_tar_out_dict["targets"]:
+            for tar in inp_tar_out_dict["targets"][key]:
+                tar = remove_left_padding(tar, tokenizer)[0]
+                llm_eval_examples["targets"].append(tokenizer.decode(tar))
```
Comment (justinxzhao) [inline ludwig/trainers/trainer_llm.py:453]: Could we do `tokenizer.decode(tar, skip_special_symbols=True)` to avoid manually removing left padding?

File: ludwig/contribs/mlflow/__init__.py [hunk 3]
```diff
@@ -241,7 +249,15 @@ def _log_mlflow(log_metrics, steps, save_path, should_continue, log_artifacts: b
 
     This is used when save_in_background is False.
     """
+    # breakpoint()
     if log_metrics is not None:
+        if log_metrics["llm_eval_examples"] is not None:
+            mlflow.llm.log_predictions(
+                inputs=log_metrics["llm_eval_examples"]["inputs"],
+                prompts=[0] * len(log_metrics["llm_eval_examples"]["inputs"]),
```
Comment (justinxzhao) [inline ludwig/contribs/mlflow/__init__.py:257]: Based on the example in https://mlflow.org/docs/latest/python_api/mlflow.llm.html, perhaps we should make `inputs=[0] * len(log_metrics["llm_eval_examples"]["inputs"])`  and `prompts=log_metrics["llm_eval_examples"]["inputs"]`. WDYT?

File: ludwig/contribs/mlflow/__init__.py [hunk 4]
```diff
@@ -241,7 +249,15 @@ def _log_mlflow(log_metrics, steps, save_path, should_continue, log_artifacts: b
...
+                outputs=log_metrics["llm_eval_examples"]["outputs"],
+            )
+            del log_metrics["llm_eval_examples"]
```
Comment (justinxzhao) [inline ludwig/contribs/mlflow/__init__.py:260]: Why is this necessary?

File: ludwig/contribs/mlflow/__init__.py [hunk 5]
```diff
@@ -225,6 +225,14 @@ def _log_mlflow_loop(q: queue.Queue, log_artifacts: bool = True):
...
             break
 
+        if log_metrics is not None:
+            if log_metrics["llm_eval_examples"] is not None:
+                mlflow.llm.log_predictions(
+                    inputs=log_metrics["llm_eval_examples"]["inputs"],
+                    prompts=log_metrics["llm_eval_examples"]["targets"],
+                    outputs=log_metrics["llm_eval_examples"]["outputs"],
+                )
+                del log_metrics["llm_eval_examples"]
```
Comment (justinxzhao) [inline ludwig/contribs/mlflow/__init__.py:235]: Why is this necessary?

File: ludwig/contribs/mlflow/__init__.py [hunk 6]
```diff
@@ -241,7 +249,15 @@ def _log_mlflow(log_metrics, steps, save_path, should_continue, log_artifacts: b
 
     This is used when save_in_background is False.
     """
+    # breakpoint()
```
Comment (justinxzhao) [inline ludwig/contribs/mlflow/__init__.py:252]: nit: Remove breakpoint.

File: ludwig/models/predictor.py [hunk 7]
```diff
@@ -388,6 +390,13 @@ def batch_evaluation(self, dataset, collect_predictions=False, collect_logits=Fa
                     outputs = self._predict_on_inputs((inputs, targets))
                     preds = self.model.outputs_to_predictions(outputs)
 
+                    for key in inputs:
+                        example_inputs[key].extend(inputs[key])
+                    for key in targets:
+                        example_targets[key].extend(targets[key])
+                    for key in preds:
+                        example_outputs[key].extend(preds[key]["predictions"])
```
Comment (justinxzhao) [inline ludwig/models/predictor.py:398]: Is there a `response` key from the predictions dictionary? I remember adding it for computing decoded-text metrics.

File: ludwig/contribs/mlflow/__init__.py (hunk 3 above)
Comment (Infernaught) [reply ludwig/contribs/mlflow/__init__.py:257]: Oh wait this is an error. Prompts should actually contain the targets (as per discussion with @geoffreyangus). Let me fix this.

File: ludwig/contribs/mlflow/__init__.py (hunk 5 above)
Comment (Infernaught) [reply ludwig/contribs/mlflow/__init__.py:235]: Are you referring to `del log_metrics["llm_eval_examples"]`? This is so that mlflow doesn't try to log the llm_eval_examples in `mlflow.log_metrics`, which would cause an error.

File: ludwig/models/predictor.py (hunk 7 above)
Comment (Infernaught) [reply ludwig/models/predictor.py:398]: In `preds[key]`, I only see "predictions", "probabilities", and "logits"

File: ludwig/models/predictor.py [hunk 8]
```diff
@@ -351,7 +351,9 @@ def batch_evaluation(self, dataset, collect_predictions=False, collect_logits=Fa
         """
```
Comment (justinxzhao) [inline ludwig/models/predictor.py:351]: Update the docstring for `batch_evaluation()`.

File: ludwig/contribs/mlflow/__init__.py (hunk 5 above)
Comment (justinxzhao) [reply ludwig/contribs/mlflow/__init__.py:235]: Makes sense. Could you add a comment about this in both places?

File: ludwig/utils/trainer_utils.py [hunk 9]
```diff
@@ -179,6 +180,7 @@ def __init__(
         self.train_metrics = train_metrics
         self.validation_metrics = validation_metrics
         self.test_metrics = test_metrics
+        self.llm_eval_examples = llm_eval_examples
```
Comment (justinxzhao) [inline ludwig/utils/trainer_utils.py:183]: Can you add a comment here with an example of the expected structure and how this is only set for LLM fine-tuning?

> llm_eval_examples is used to capture samples of prompts/outputs for LLM fine-tuning. There are 3 keys (inputs, targets, outputs), which each map to a list of strings.

File: ludwig/utils/trainer_utils.py [hunk 10]
```diff
@@ -210,6 +212,7 @@ def log_metrics(self):
             "best_valid_metric": self.best_eval_metric_value,
             "num_reductions_lr": self.num_reductions_learning_rate,
             "num_increases_bs": self.num_increases_batch_size,
+            "llm_eval_examples": self.llm_eval_examples,
```
Comment (justinxzhao) [inline ludwig/utils/trainer_utils.py:215]: Maybe add a comment here too about how the structure of llm_eval_examples is different from all other entries and certain callbacks.

> Note that `llm_eval_examples` is the only entry in the map with string leaves. Loggers and other callbacks should handle this accordingly.

Comment (justinxzhao) [inline ludwig/utils/trainer_utils.py:215]: nit: How about:

```python
if self.llm_eval_examples:
    log_metrics["llm_eval_examples"] = self.llm_eval_examples
```

**Review by justinxzhao — COMMENTED** [review]
Thanks for the changes @Infernaught! LGTM after adding a few additional comments.

**Comment by justinxzhao** [comment]
If you rebase, the CI should run clean thanks to @alexsherstinsky's recent changes.

File: ludwig/trainers/trainer.py [hunk 11]
```diff
@@ -638,6 +638,7 @@ def run_evaluation(
         all_losses: Dict[str, torch.Tensor],
         early_stopping_steps: int,
         checkpoint_manager: CheckpointManager,
+        num_eval_examples: int = 5,
```
Comment (justinxzhao) [inline ludwig/trainers/trainer.py:641]: Is this being used anywhere anymore?

**Review by justinxzhao — APPROVED** [review]

File: ludwig/models/predictor.py [hunk 12]
```diff
@@ -416,9 +427,10 @@ def batch_evaluation(self, dataset, collect_predictions=False, collect_logits=Fa
             metrics = self.model.get_metrics()
             self.model.reset_metrics()
 
-            self.dist_model.train(prev_model_training_mode)  # Restores previous model training mode.
+            inp_tar_out_dict = {"inputs": example_inputs, "targets": example_targets, "outputs": example_outputs}
```
Comment (arnavgarg1) [inline ludwig/models/predictor.py:430]: nit, can we rename this to `input_target_output_dict` for clarity?
